File: backend/satellites.py
"""
卫星目录 REST 接口：列表 / 搜索 / 导入 / 删除 / 详情 / 介绍频率 / 手动刷新。
"""

from __future__ import annotations

import time
from datetime import datetime, timezone

from fastapi import APIRouter

from exceptions import NotFoundError, ValidationError  # noqa: E402
from provider import fetch_satellite_info_online, fetch_tle_online  # noqa: E402
from store import (  # noqa: E402
    _COMMON_SATELLITES,
    _get_amsat_freq_map,
    _load_sat_info,
    _load_satellites,
    _load_tles,
    _save_sat_info,
    _save_satellites,
    _save_tle,
)
from tle import (  # noqa: E402
    _INFO_VALID_SECONDS,
    _TLE_VALID_SECONDS,
    _clean_sat_name,
    _get_tle_cached,
    _parse_tle_epoch,
    _parse_tle_fields,
    tle_cache_set,
    tle_source,
)

router = APIRouter()


def _find_satellite(sat_id: str) -> dict | None:
    """按 id 或 NORAD 目录号查找卫星。"""
    sats = _load_satellites()
    return next((s for s in sats if s["id"] == sat_id or str(s["norad_id"]) == sat_id), None)


@router.get("/api/satellites")
def api_get_satellites():
    """读取卫星列表（内置 + 自定义），附带 TLE 更新时间与轨道历元。"""
    sats = _load_satellites()
    tles = _load_tles()
    out = []
    for s in sats:
        item = dict(s)
        saved = tles.get(str(s["norad_id"]))
        if saved:
            fetched_ts = float(saved["fetched_ts"])
            item["fetched_at"] = datetime.fromtimestamp(fetched_ts, tz=timezone.utc).isoformat()
            epoch = _parse_tle_epoch(saved["tle1"])
            item["epoch"] = epoch.isoformat() if epoch else ""
        else:
            item["fetched_at"] = ""
            item["epoch"] = ""
        out.append(item)
    return {"satellites": out}


@router.post("/api/satellites/search")
def api_search_satellites(payload: dict):
    """按名称或 NORAD 目录号搜索卫星（供导入前选择）。

    名称搜索：在常用卫星目录中做中英文模糊匹配，再在线获取 TLE；
    纯数字：直接按 NORAD 目录号查询。
    """
    query = str(payload.get("query", "")).strip()
    if not query:
        raise ValidationError("请输入卫星名称或 NORAD 目录号")
    # 纯数字 → 直接按 NORAD 目录号查询
    if query.isdigit():
        nid = int(query)
        tle = fetch_tle_online(norad_id=nid)
        if tle is None:
            raise NotFoundError(f"未找到 NORAD {nid} 对应的卫星")
        tle_name, tle1, tle2 = tle
        return {
            "results": [
                {
                    "name": _clean_sat_name(tle_name),
                    "norad_id": nid,
                    "tle1": tle1,
                    "tle2": tle2,
                }
            ]
        }
    # 名称搜索：在常用卫星目录中匹配（中英文，不区分大小写），即时返回候选列表
    # TLE 在用户确认导入时再在线获取，避免多颗匹配时等待过久
    q = query.lower()
    matched = [
        s for s in _COMMON_SATELLITES
        if q in s["name"].lower() or any(q in a.lower() for a in s.get("aliases", []))
    ]
    if not matched:
        raise NotFoundError(f"未找到与“{query}”匹配的卫星，可尝试英文名称或 NORAD 目录号")
    return {"results": [{"name": s["name"], "norad_id": s["norad_id"]} for s in matched[:8]]}


@router.post("/api/satellites/import")
def api_import_satellite(payload: dict):
    """按 NORAD 目录号从网络导入卫星：联网获取 TLE 验证成功后加入列表。"""
    norad = str(payload.get("norad_id", "")).strip()
    if not norad.isdigit():
        raise ValidationError("请输入有效的 NORAD 目录号（纯数字）")
    nid = int(norad)
    sats = _load_satellites()
    if any(int(s["norad_id"]) == nid for s in sats):
        raise ValidationError(f"NORAD {nid} 已在卫星列表中")
    try:
        tle = fetch_tle_online(norad_id=nid)
    except Exception as exc:
        raise ValidationError(f"获取 TLE 失败：{exc}") from exc
    if tle is None:
        raise NotFoundError(f"未能在网络上找到 NORAD {nid} 的有效 TLE（请检查目录号）")
    name, _l1, _l2 = tle
    name = _clean_sat_name(name)  # 清理前导编号（如 "0 NOAA 19" → "NOAA 19"）
    sat = {"id": str(nid), "name": name, "norad_id": nid, "builtin": False}
    sats.append(sat)
    saved = _save_satellites(sats)
    return {"satellites": saved, "satellite": sat}


@router.post("/api/satellites/delete")
def api_delete_satellite(payload: dict):
    """删除自定义卫星（内置卫星不可删除）。"""
    sid = str(payload.get("id", ""))
    sats = [s for s in _load_satellites() if not (s["id"] == sid and not s["builtin"])]
    saved = _save_satellites(sats)
    return {"satellites": saved}


@router.get("/api/satellites/{sat_id}")
def api_satellite_detail(sat_id: str):
    """卫星详情：基本信息 + 最新 TLE + 解析出的轨道根数 + 数据时间/过期状态。"""
    sat = _find_satellite(sat_id)
    if not sat:
        raise NotFoundError("卫星不存在")
    nid = int(sat["norad_id"])
    tle_name, tle1, tle2 = _get_tle_cached(nid)
    saved = _load_tles().get(str(nid))
    fetched_ts = float(saved["fetched_ts"]) if saved else None
    return {
        "id": sat["id"],
        "name": sat["name"],
        "norad_id": nid,
        "builtin": sat["builtin"],
        "tle_name": tle_name,
        "tle1": tle1,
        "tle2": tle2,
        "fetched_at": datetime.fromtimestamp(fetched_ts, tz=timezone.utc).isoformat() if fetched_ts else "",
        "tle_age_hours": round((time.time() - fetched_ts) / 3600, 1) if fetched_ts else None,
        "tle_stale": bool(fetched_ts and (time.time() - fetched_ts) > _TLE_VALID_SECONDS),
        "tle_source": tle_source(nid),
        "orbit": _parse_tle_fields(tle1, tle2),
    }


@router.get("/api/satellites/{sat_id}/info")
def api_satellite_info(sat_id: str):
    """卫星介绍与频率信息：SatNOGS 基本信息 + AMSAT 业余卫星频率（30 天本地缓存）。"""
    sat = _find_satellite(sat_id)
    if not sat:
        raise NotFoundError("卫星不存在")
    nid = int(sat["norad_id"])
    now = time.time()
    infos = _load_sat_info()
    saved = infos.get(str(nid))
    if saved and (now - float(saved.get("fetched_ts", 0))) < _INFO_VALID_SECONDS:
        info = saved
    else:
        meta = fetch_satellite_info_online(nid) or {}
        freqs = _get_amsat_freq_map().get(str(nid), [])
        info = {
            "names": meta.get("names", ""),
            "status": meta.get("status", ""),
            "launch_date": meta.get("launch_date", ""),
            "operator": meta.get("operator", ""),
            "countries": meta.get("countries", ""),
            "website": meta.get("website", ""),
            "telemetries": meta.get("telemetries", []),
            "frequencies": freqs,
            "norad_id": nid,
            "fetched_ts": now,
        }
        _save_sat_info(nid, info, now)
    return {
        "id": sat["id"],
        "name": sat["name"],
        "norad_id": nid,
        "names": info.get("names", ""),
        "status": info.get("status", ""),
        "launch_date": info.get("launch_date", ""),
        "operator": info.get("operator", ""),
        "countries": info.get("countries", ""),
        "website": info.get("website", ""),
        "telemetries": info.get("telemetries", []),
        "frequencies": info.get("frequencies", []),
        "fetched_at": datetime.fromtimestamp(float(info.get("fetched_ts", now)), tz=timezone.utc).isoformat(),
    }


@router.post("/api/satellites/{sat_id}/refresh")
def api_refresh_satellite(sat_id: str):
    """强制从网络刷新指定卫星的 TLE 并持久化（手动更新轨道数据）。"""
    sat = _find_satellite(sat_id)
    if not sat:
        raise NotFoundError("卫星不存在")
    nid = int(sat["norad_id"])
    try:
        tle = fetch_tle_online(norad_id=nid)
    except Exception as exc:
        raise ValidationError(f"在线刷新失败：{exc}") from exc
    if tle is None:
        raise ValidationError(f"在线刷新失败：未获取到 NORAD {nid} 的有效 TLE（请稍后重试）")
    tle_name, tle1, tle2 = tle
    name = _clean_sat_name(tle_name)
    now = time.time()
    _save_tle(nid, name, tle1, tle2, now, source="online")
    tle_cache_set(nid, name, tle1, tle2, now, source="online")  # 同步内存缓存
    return {
        "id": sat["id"],
        "name": name,
        "norad_id": nid,
        "builtin": sat["builtin"],
        "tle_name": name,
        "tle1": tle1,
        "tle2": tle2,
        "fetched_at": datetime.fromtimestamp(now, tz=timezone.utc).isoformat(),
        "tle_age_hours": 0.0,
        "tle_stale": False,
        "tle_source": "online",
        "orbit": _parse_tle_fields(tle1, tle2),
    }


@router.post("/api/satellites/refresh-all")
def api_refresh_all_satellites():
    """批量从网络更新全部卫星的 TLE（逐颗获取并持久化，失败不影响其它卫星）。"""
    sats = _load_satellites()
    results = []
    for sat in sats:
        nid = int(sat["norad_id"])
        try:
            tle = fetch_tle_online(norad_id=nid)
        except Exception as exc:
            results.append({"id": sat["id"], "norad_id": nid, "ok": False, "error": str(exc)})
            continue
        if tle is None:
            results.append({"id": sat["id"], "norad_id": nid, "ok": False, "error": "在线获取 TLE 失败"})
            continue
        tle_name, tle1, tle2 = tle
        name = _clean_sat_name(tle_name)
        now = time.time()
        _save_tle(nid, name, tle1, tle2, now, source="online")
        tle_cache_set(nid, name, tle1, tle2, now, source="online")  # 同步内存缓存
        results.append(
            {
                "id": sat["id"],
                "norad_id": nid,
                "ok": True,
                "fetched_at": datetime.fromtimestamp(now, tz=timezone.utc).isoformat(),
            }
        )
    updated = sum(1 for r in results if r["ok"])
    failed = len(results) - updated
    return {"results": results, "updated": updated, "failed": failed}

File: backend-rs/src/routes/satellites.rs
//! 卫星目录 REST 接口：对应 backend/satellites.py。
//!
//! 列表 / 搜索 / 导入 / 删除 / 详情 / 介绍频率 / 手动刷新。

use axum::Json;
use axum::extract::Path;
use serde_json::{Value, json};

use crate::astroconv::ts_to_iso;
use crate::error::{ApiError, ApiResult};
use crate::libfiles;
use crate::provider;
use crate::store;
use crate::tle;

const TLE_VALID_SECONDS: f64 = 12.0 * 3600.0;

/// 按 id 或 NORAD 目录号查找卫星。
fn find_satellite(sat_id: &str) -> Option<Value> {
    store::load_satellites().into_iter().find(|s| {
        let id_match = s.get("id").and_then(|v| v.as_str()) == Some(sat_id);
        let norad_match = s
            .get("norad_id")
            .and_then(|v| v.as_i64())
            .map(|n| n.to_string())
            .as_deref()
            == Some(sat_id);
        id_match || norad_match
    })
}

/// 从请求体读取 norad_id（兼容字符串与数字），返回其字符串形式。
fn norad_from_payload(payload: &Value) -> String {
    match payload.get("norad_id") {
        Some(Value::String(s)) => s.trim().to_string(),
        Some(Value::Number(n)) => n
            .as_i64()
            .map(|i| i.to_string())
            .unwrap_or_else(|| n.to_string()),
        _ => String::new(),
    }
}

/// 计算 TLE 获取时间戳（无则 None）。
fn fetched_ts_for(norad_id: i64) -> Option<f64> {
    store::load_tles()
        .get(&norad_id.to_string())
        .and_then(|v| v.get("fetched_ts"))
        .and_then(|v| v.as_f64())
}

/// GET /api/satellites：卫星列表（内置 + 自定义），附 TLE 更新时间与轨道历元。
pub async fn get_satellites() -> Json<Value> {
    let sats = store::load_satellites();
    let tles = store::load_tles();
    let out: Vec<Value> = sats
        .iter()
        .map(|s| {
            let mut item = s.clone();
            let norad = s.get("norad_id").and_then(|v| v.as_i64()).unwrap_or(0);
            match tles.get(&norad.to_string()) {
                Some(saved) => {
                    let fetched = saved.get("fetched_ts").and_then(|v| v.as_f64()).unwrap_or(0.0);
                    item["fetched_at"] = json!(ts_to_iso(fetched));
                    let tle1 = saved.get("tle1").and_then(|v| v.as_str()).unwrap_or("");
                    item["epoch"] = json!(
                        tle::parse_tle_epoch(tle1)
                            .map(|d| crate::astroconv::iso_utc(&d))
                            .unwrap_or_default()
                    );
                }
                None => {
                    item["fetched_at"] = json!("");
                    item["epoch"] = json!("");
                }
            }
            item
        })
        .collect();
    Json(json!({ "satellites": out }))
}

/// POST /api/satellites/import：按 NORAD 目录号从网络导入卫星。
pub async fn import_satellite(Json(payload): Json<Value>) -> ApiResult<Json<Value>> {
    let norad = norad_from_payload(&payload);
    if norad.is_empty() || !norad.chars().all(|c| c.is_ascii_digit()) {
        return Err(ApiError::validation("请输入有效的 NORAD 目录号（纯数字）"));
    }
    let nid: i64 = norad.parse().unwrap_or(0);
    let sats = store::load_satellites();
    if sats
        .iter()
        .any(|s| s.get("norad_id").and_then(|v| v.as_i64()) == Some(nid))
    {
        return Err(ApiError::validation(format!("NORAD {} 已在卫星列表中", nid)));
    }
    match provider::fetch_tle_online(nid).await {
        None => Err(ApiError::not_found(format!(
            "未能在网络上找到 NORAD {} 的有效 TLE（请检查目录号）",
            nid
        ))),
        Some((name, _l1, _l2)) => {
            let name = tle::clean_sat_name(&name);
            let sat = json!({ "id": norad, "name": name, "norad_id": nid, "builtin": false });
            let mut next = sats.clone();
            next.push(sat.clone());
            let saved = store::save_satellites(&Value::Array(next));
            Ok(Json(json!({ "satellites": saved, "satellite": sat })))
        }
    }
}

/// POST /api/satellites/delete：删除自定义卫星（内置不可删除）。
pub async fn delete_satellite(Json(payload): Json<Value>) -> Json<Value> {
    let sid = payload
        .get("id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let sats: Vec<Value> = store::load_satellites()
        .into_iter()
        .filter(|s| {
            let is_target = s.get("id").and_then(|v| v.as_str()) == Some(sid.as_str());
            let builtin = s.get("builtin").and_then(|v| v.as_bool()).unwrap_or(false);
            !(is_target && !builtin)
        })
        .collect();
    let saved = store::save_satellites(&Value::Array(sats));
    Json(json!({ "satellites": saved }))
}

/// GET /api/satellites/{sat_id}：卫星详情 + 最新 TLE + 轨道根数 + 数据时间/过期状态。
pub async fn satellite_detail(Path(sat_id): Path<String>) -> ApiResult<Json<Value>> {
    let sat = find_satellite(&sat_id).ok_or_else(|| ApiError::not_found("卫星不存在"))?;
    let nid = sat.get("norad_id").and_then(|v| v.as_i64()).unwrap_or(0);
    let (tle_name, tle1, tle2) = tle::get_tle_cached(nid)
        .await
        .map_err(|e| {
            tracing::error!("获取 TLE 失败: {}", e);
            ApiError::service("服务器内部错误")
        })?;
    let fetched_ts = fetched_ts_for(nid);
    let now = store::now_ts();
    let orbit = tle::parse_tle_fields(&tle1, &tle2);
    Ok(Json(json!({
        "id": sat.get("id").cloned().unwrap_or(Value::Null),
        "name": sat.get("name").cloned().unwrap_or(Value::Null),
        "norad_id": nid,
        "builtin": sat.get("builtin").cloned().unwrap_or(Value::Null),
        "tle_name": tle_name,
        "tle1": tle1,
        "tle2": tle2,
        "fetched_at": fetched_ts.map(ts_to_iso).unwrap_or_default(),
        "tle_age_hours": fetched_ts.map(|ts| json!((crate::astro::round((now - ts) / 3600.0, 1)))).unwrap_or(Value::Null),
        "tle_stale": fetched_ts.map(|ts| (now - ts) > TLE_VALID_SECONDS).unwrap_or(false),
        "tle_source": tle::tle_source(nid),
        "orbit": orbit,
    })))
}

/// GET /api/satellites/{sat_id}/info：卫星介绍与频率（SatNOGS + AMSAT）。
pub async fn satellite_info(Path(sat_id): Path<String>) -> ApiResult<Json<Value>> {
    let sat = find_satellite(&sat_id).ok_or_else(|| ApiError::not_found("卫星不存在"))?;
    let nid = sat.get("norad_id").and_then(|v| v.as_i64()).unwrap_or(0);
    let mut base = json!({
        "id": sat.get("id").cloned().unwrap_or(Value::Null),
        "name": sat.get("name").cloned().unwrap_or(Value::Null),
        "norad_id": nid,
        "names": "", "status": "", "launch_date": "", "operator": "",
        "countries": "", "website": "", "telemetries": [], "frequencies": [],
        "fetched_at": "",
    });
    if let Some(info) = libfiles::get_satellite_info(nid, false).await {
        if let (Some(b), Some(i)) = (base.as_object_mut(), info.as_object()) {
            for (k, v) in i {
                b.insert(k.clone(), v.clone());
            }
        }
    }
    Ok(Json(base))
}

/// POST /api/satellites/{sat_id}/refresh：强制从网络刷新指定卫星的 TLE。
pub async fn refresh_satellite(Path(sat_id): Path<String>) -> ApiResult<Json<Value>> {
    let sat = find_satellite(&sat_id).ok_or_else(|| ApiError::not_found("卫星不存在"))?;
    let nid = sat.get("norad_id").and_then(|v| v.as_i64()).unwrap_or(0);
    let (tle_name, tle1, tle2) = provider::fetch_tle_online(nid).await.ok_or_else(|| {
        ApiError::validation(format!(
            "在线刷新失败：未获取到 NORAD {} 的有效 TLE（请稍后重试）",
            nid
        ))
    })?;
    let name = tle::clean_sat_name(&tle_name);
    let now = store::now_ts();
    store::save_tle(nid, &name, &tle1, &tle2, now, "online");
    tle::tle_cache_set(nid, &name, &tle1, &tle2, now, "online");
    let orbit = tle::parse_tle_fields(&tle1, &tle2);
    Ok(Json(json!({
        "id": sat.get("id").cloned().unwrap_or(Value::Null),
        "name": name,
        "norad_id": nid,
        "builtin": sat.get("builtin").cloned().unwrap_or(Value::Null),
        "tle_name": name,
        "tle1": tle1,
        "tle2": tle2,
        "fetched_at": ts_to_iso(now),
        "tle_age_hours": 0.0,
        "tle_stale": false,
        "tle_source": "online",
        "orbit": orbit,
    })))
}

/// POST /api/satellites/refresh-all：批量从网络更新全部卫星的 TLE。
pub async fn refresh_all_satellites() -> Json<Value> {
    let sats = store::load_satellites();
    let mut results: Vec<Value> = Vec::new();
    for sat in sats {
        let nid = sat.get("norad_id").and_then(|v| v.as_i64()).unwrap_or(0);
        let id = sat.get("id").cloned().unwrap_or(Value::Null);
        match provider::fetch_tle_online(nid).await {
            None => results.push(json!({
                "id": id, "norad_id": nid, "ok": false, "error": "在线获取 TLE 失败",
            })),
            Some((tle_name, tle1, tle2)) => {
                let name = tle::clean_sat_name(&tle_name);
                let now = store::now_ts();
                store::save_tle(nid, &name, &tle1, &tle2, now, "online");
                tle::tle_cache_set(nid, &name, &tle1, &tle2, now, "online");
                results.push(json!({
                    "id": id, "norad_id": nid, "ok": true, "fetched_at": ts_to_iso(now),
                }));
            }
        }
    }
    let updated = results
        .iter()
        .filter(|r| r.get("ok").and_then(|v| v.as_bool()).unwrap_or(false))
        .count();
    let failed = results.len() - updated;
    Json(json!({ "results": results, "updated": updated, "failed": failed }))
}
